### ISSUE TITLE
[Common Lisp] Implement new Concept Exercise: code-as-data

### ISSUE THREAD

**Issue by TheLostLambda** [exercism/v3#1828]
This issue describes how to implement the `code-as-data` concept exercise for the Common Lisp track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Learning objectives

- Understand, in layman's terms, what "homoiconicity" means for Lisp (lists can be evaluated as code)
- Be able to define new, simple macros with `defmacro`
- Know the difference between compile-time / load-time and run-time evaluation in the context of `defmacro` vs `defun`
- Be aware of the additional differences between `defmacro` and `defun` (arguments aren't evaluated, macros can't be `map`ped, `funcall`ed, or `apply`d)
- Understand quasi-quoting and how it can be used to simplify list construction (```, `,`, and `,@`)
- Be able to explore / debug macros with `macroexpand`

## Out of scope

- Compiler macros
- Reader macros
- Symbol macros
- Hygienic macros with `gensym`
- Using macros for optimisation (performing calculations at compile-time)

## Concepts

- `code-as-data`

## Prerequisites

- `basic-lists`
- `arithmetic`
- `symbols`
- `function-definition`
- `expressions`
- `enumeration`
- `higher-order-functions`

### Hints

- [CL Cookbook](https://lispcookbook.github.io/cl-cookbook/macros.html)

### After

- [2 Hours of Macros (YouTube)](https://www.youtube.com/watch?v=ygKXeLKhiTI)
- [CL-PPCRE](https://edicl.github.io/cl-ppcre/) uses compiler macros to translate many regular expressions into normal lisp code, leading to a very high-performance regex library – it's all just sugar for s-expressions.

## Representer

We may need to double check that "variable-like" symbols being defined in macros are properly replaced with a name mapping

## Analyzer

Nothing yet, should be able to use `macroexpand` to help with tests

## Implementing

While it's important that the student knows how to debug macros using things like `macroexpand`, that's not something we need to test for.

This one might be one of the trickier concepts, even if the functions and syntax involved aren't that complex. It's vital that the student understands what macros are good for and what they aren't, so that they don't over-apply them. Convincing examples are important here, even if they aren't super trivial. See ["What Macros are For"](https://lispcookbook.github.io/cl-cookbook/macros.html#what-macros-are-for) for an example.

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.
